tiledb/api/src/query/read/output/arrow.rs

```diff
@@ -0,0 +1,340 @@
+            .collect::<Vec<i64>>();
+
+        let field = Arc::new(Field::new_list_field(
+            arrow_values.data_type().clone(),
+            value.input.validity.is_some(),
+        ));
+
+        let arrow_offsets =
+            OffsetBuffer::<i64>::new(ScalarBuffer::<i64>::from(v_offsets));
+
+        let arrow_validity = if let Some(validity) = value.input.validity {
+            let validity: Vec<u8> = match validity {
+                Buffer::Empty => vec![],
+                Buffer::Owned(validity) => validity.into_vec(),
+                Buffer::Borrowed(validity) => {
+                    validity[0..value.nvalues].to_vec()
+                }
+            };
+            let arrow_validity = validity
+                .into_iter()
+                .take(value.nvalues)
+                .map(|v: u8| v != 0)
+                .collect::<arrow::buffer::NullBuffer>();
+            Some(arrow_validity)
+        } else {
+            None
+        };
+
+        let list_array = GenericListArray::<i64>::try_new(
+            field,
+            arrow_offsets,
+            Arc::new(arrow_values),
+            arrow_validity,
+        )
+        .expect("TileDB internal error constructing Arrow buffers");
+
+        Arc::new(list_array)
+    }
+}
+
+impl From<TypedRawReadOutput<'_>> for Arc<dyn ArrowArray> {
+    fn from(value: TypedRawReadOutput<'_>) -> Self {
+        typed_query_buffers_go!(value.buffers, _DT, input, {
+            RawReadOutput {
+                nvalues: value.nvalues,
+                nbytes: value.nbytes,
+                input,
+            }
+            .into()
+        })
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::any::TypeId;
+
+    use super::*;
+    use proptest::prelude::*;
+
+    fn raw_read_to_arrow<C>(rr: RawReadOutput<C>)
+    where
+        C: ArrowPrimitiveTypeNative,
+        PrimitiveArray<<C as ArrowPrimitiveTypeNative>::ArrowPrimitiveType>:
+            From<Vec<C>> + From<Vec<Option<C>>>,
+    {
+        type MyPrimitiveArray<C> =
+            PrimitiveArray<<C as ArrowPrimitiveTypeNative>::ArrowPrimitiveType>;
+
+        let arrow = {
+            let rrborrow = RawReadOutput {
+                nvalues: rr.nvalues,
+                nbytes: rr.nbytes,
+                input: rr.input.borrow(),
+            };
+            Arc::<dyn ArrowArray>::from(rrborrow)
+        };
+
+        if let Some(offsets) = rr.input.cell_offsets {
+            assert_eq!(
+                TypeId::of::<GenericListArray<i64>>(),
+                arrow.as_any().type_id()
+            );
+            let gl = arrow
+                .as_any()
+                .downcast_ref::<GenericListArray<i64>>()
+                .unwrap();
+
+            let arrow_offsets = gl.offsets();
+            assert!(arrow_offsets.len() <= offsets.len() + 1);
+            assert_eq!(arrow_offsets.len(), rr.nvalues + 1);
+
+            /* arrow offsets are value, tiledb offsets are bytes */
```

Comment:
Ooof. I can't wait to deal with all the bugs this difference is gonna cause.

Reply:
The good news is that Isaiah said we can configure TileDB to produce offsets which are records, not bytes. Might not even be a bad plan to do that as the default.